request-example.go

```diff
@@ -37,7 +37,8 @@ func (fs *root) Fileread(r *Request) (io.ReaderAt, error) {
 		return nil, os.ErrInvalid
 	}
 
-	return fs.OpenFile(r)
+	// Needs to be readable by the user.
```

Comment:
I'm making some assumptions in all of these, and they are correct now, but wouldn't be if we expanded our supported features a bit more.

Specifically, memfile does not implement file ownership.

So I'm treating the logged in user as the owner of everything, because that follows the principal of least surprise when we don't have file ownership implement.


The main file permissions bits that we have are owner, group, and other.  (The results when other has broader permissions that owner are delightfully interesting.)

And so all of the file permissions checks are made against the owner bits.

For v2, I'm very tempted to say that the sane choice is to just us the io/fs.FS interface, and the related interfaces specified by io/fs.

At that point the package can either use an existing memory backed implementation, or if a better one is needed, it can live in a separate package and be useful outside of just sftp.

Reply:
This is strictly just a pedantic comment about `owner` vs `user` vocabulary from the standards (https://pubs.opengroup.org/onlinepubs/9699919799/): 

> … the requested access permission bit is set for the class (file owner class, file group class, or file other class)…

The fix is merely to use the word `owner` not `user` here. In order to align on the underlying standards.

Yes, for v2, we would want to align on `io/fs.FS` we also do not need to implement ownership with this PR, as we’re only adding permissions, not ownership.

Reply:
Hm, very silly question...

How hard do you think it would be to write a wrapper that takes a suitable io/fs.FS object, and implements all of the existing handlers?

That shouldn't have any externally visible API impact, except for exposing the type that does the wrapping.

Reply:
I’ve really only thought about this from the client side.

It’s sadly a lot more work than one would first think, because this package was original designed well before the `fs.FS` paradigm (or even the `ExtensionInterface` paradigm) were released, not even widespread.

I really do fear that it would be just as complex of a task as refactoring and restructuring the code at least fairly deeply. And then we would be left with three server implementations: `Server`, `RequestServer`, and `FSServer`. :( Working on a `v2` keeps getting more and more pressing, but you know F/OSS, it can be hard to find priority to put into a project, unless someone doing it as their job is tasked with helping out.

Reply:
I was actually thinking of implementing it at the same level as Memfile, so you would have RequestServer, and your request handler could be a wrapper for fs.FS.